sdk/src/main/java/com/hedera/hashgraph/sdk/TopicMessageQuery.java

```diff
@@ -1,29 +1,43 @@
 package com.hedera.hashgraph.sdk;
 
-import com.hedera.hashgraph.sdk.proto.TransactionID;
-import com.hedera.hashgraph.sdk.proto.mirror.ConsensusServiceGrpc;
-import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicQuery;
-import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicResponse;
 import io.grpc.CallOptions;
 import io.grpc.ClientCall;
 import io.grpc.Status;
 import io.grpc.StatusRuntimeException;
 import io.grpc.stub.ClientCalls;
 import io.grpc.stub.StreamObserver;
-import java8.util.function.BiConsumer;
-import java8.util.function.Consumer;
-import org.threeten.bp.Instant;
-
-import javax.annotation.Nullable;
+import java.time.Duration;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.threeten.bp.Instant;
+
+import com.hedera.hashgraph.sdk.proto.TransactionID;
+import com.hedera.hashgraph.sdk.proto.mirror.ConsensusServiceGrpc;
+import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicQuery;
+import com.hedera.hashgraph.sdk.proto.mirror.ConsensusTopicResponse;
 
 public final class TopicMessageQuery {
-    private final ConsensusTopicQuery.Builder builder;
 
-    @Nullable
-    private BiConsumer<Throwable, TopicMessage> errorHandler;
+    private static final Logger LOGGER = LoggerFactory.getLogger(TopicMessageQuery.class);
+    private static final Pattern RST_STREAM = Pattern
+            .compile(".*(rst.stream.*internal.error|internal.error.*rst.stream).*",
+                    Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
+
+    private final ConsensusTopicQuery.Builder builder;
+    private Runnable completionHandler = this::onComplete;
+    private BiConsumer<Throwable, TopicMessage> errorHandler = this::onError;
```

Comment:
This could be changed to a `Consumer<Throwable>` since a non-null `TopicMessage` will never be passed to it now with the change to not call the error handle for application errors during `onNext()`. But I didn't want to break API.

Reply:
One question that comes to mind, how will the user know their code erred in the callback? We had an issue with another project awhile ago where the application code was faulty and died within the callback; the user wasn't informed of this and didn't know if it was their code, or the SDK's code that was faulty, but all the knew is the callback stopped being called.

Another thing to mention is that I know there are some project that already use this specific feature, changing this behavior will break their code at runtime when the update to the latest SDK. So, although this isn't a API breaking change, I think it's fair to say this is still a breaking change.

Reply:
> how will the user know their code erred in the callback?

That's the thing, it's their code. They should be catching the error and handling it.

But the breaking compatibility is a fair point. So I tested it and I think it doesn't break compatibility. If you throw your own exception inside the custom `onNext()` handler, it will cause the stream to cancel and wrap that user's exception in a `io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.` That exception will propagate to the `onError()`. Since our retryHandler doesn't retry status codes of `CANCELLED` by default it will in turn call the `errorHandler`. So effectively it is the same behavior of calling the `errorHandler` without explicitly calling, resulting in cleaner code. Please test it yourself to confirm, though.

Reply:
This is definitely still a breaking change that the exception is being wrapped. 

> That's the thing, it's their code. They should be catching the error and handling it.

The behavior here was modeled after how promises work in JavaScript. If you throw an exception in the resolver it calls the same error handler as if the promise initially resolved with an exception.

In languages that don't have exceptions, where does the error go? The error has to be returned from `onNext` or handled in `onNext`. In the latter, the user would always need to write a second function, that is called in `onNext` to handle their errors.